m3/src/reporter.cc

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "m3/reporter.h"
+
+#include <chrono>
+#include <map>
+#include <string>
+
+#include "m3/src/reporter_impl.h"
+#include "tally/src/capable_of.h"
+
+namespace tally {
+
+namespace m3 {
+
+Reporter::Reporter(const std::string &host, uint16_t port,
+                   const std::map<std::string, std::string> &common_tags,
+                   uint32_t max_queue_size, uint16_t max_packet_size)
+    : impl_(new Reporter::Impl(host, port, common_tags, max_queue_size,
+                               max_packet_size)) {}
+
+Reporter::~Reporter() = default;
```

Comment:
I attempted to define it in the header but since the destructor of the Impl class isn't defined yet the code fails to compile. I've added a comment explaining why we must define it in the class file (since the destructor of the Impl class has been defined at this point) as a compromise.